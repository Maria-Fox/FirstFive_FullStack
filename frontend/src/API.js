import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "https://localhost:3001";

class API {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {
    // console.debug("API Call:", endpoint, data, method, this.token, "**********");

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}`, 'Access-Control-Allow-Origin': 'https://firstfive.onrender.comtps://firstfive.surge.sh'};
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      // console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };

  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };




  // ******************************************* USER methods

  // Need to see about limmited views to only those who matched..


  static async viewUser(username) {
    let res = await this.request(`users/${username}`);
    return res.userData;
  };

  static async viewAuthUserProfile(username) {
    let res = await this.request(`users/${username}/profile`);
    return res.userData;
  };

  static async editUser(username, updateData) {
    let res = await this.request(`users/${username}`, { ...updateData }, 'patch');
    return res.userData;
  };

  static async deleteUser(username) {
    let res = await this.request(`users/${username}`, {}, 'delete');
    return res;
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };


  // ******************************************* Project Member methods

  static async addProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/add/${project_id}`, { username }, 'post');
    return res;
  };

  static async viewAllProjMembers(project_id) {
    let res = await this.request(`projectmembers/${project_id}/users`);
    return res;
  };

  static async deleteProjectMember(project_id, username) {
    let res = await this.request(`projectmembers/${project_id}`, { username }, 'delete');
    return res;
  };

  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };


  // class closing bracket don't delete.
};

export default API;