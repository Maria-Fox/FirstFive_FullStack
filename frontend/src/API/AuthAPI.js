import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class AuthAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Register/ LOGIN methods


  static async registerUser(userData) {
    let newUser = await this.request("auth/register", { ...userData }, 'post');
    this.token = newUser.signedJWT
    return newUser.signedJWT;
  };

  static async authenticateUser(userData) {
    let authUser = await this.request(`auth/login`, { ...userData }, 'post');
    return authUser.signedJWT;
  };

};

export default AuthAPI;