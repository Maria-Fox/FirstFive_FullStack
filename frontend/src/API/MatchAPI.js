import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MatchAPI {

  // Used to authenticate user. Sent in request header.
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Match methods

  static async addMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'post');
    return res;
  };

  static async viewUserMatches(username) {
    let res = await this.request(`matches/${username}/all`);
    return res;
  };

  static async removeUserMatch(username, project_id) {
    let res = await this.request(`matches/${username}/${project_id}`, {}, 'delete');
    return res;
  };

  static async viewProjectUserMatches(project_id) {
    let res = await this.request(`matches/${project_id}/users`);
    return res;
  };


};

export default MatchAPI;