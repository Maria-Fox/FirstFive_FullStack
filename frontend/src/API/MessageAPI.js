import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class MessageAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* Message methods

  static async createMessage(username, msgData) {
    let res = await this.request(`messages/${username}/create`, { ...msgData }, 'post');
    return res;
  };

  static async getAllUserMessages(username) {
    let res = await this.request(`messages/${username}/all`);
    return res;
  };

  static async viewSingleMsgData(message_id, username) {
    let res = await this.request(`messages/${username}/read/${message_id}`);
    return res;
  };

};

export default MessageAPI;