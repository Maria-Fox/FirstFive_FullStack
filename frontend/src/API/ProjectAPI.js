import axios from "axios";

const BASE_URL = process.env.REACT_APP_BASE_URL || "http://localhost:3001";

class ProjectAPI {

  // Used to authenticate user
  static token = "";

  // ******************************************* API send request method

  static async request(endpoint, data = {}, method = "get") {

    if (!this.token) {
      this.token = window.localStorage.getItem("token") || "invalidTemp"
    };

    const url = `${BASE_URL}/${endpoint}`;
    const headers = { Authorization: `Bearer ${API.token}` };
    const params = (method === "get")
      ? data
      : {};


    try {
      return (await axios({ url, method, data, params, headers })).data;
    } catch (err) {
      console.error("API Error:", err.message);
      // grab just the error message
      let message = err.response.data.error.message;
      throw Array.isArray(message) ? message : [message];
    };
  };


  // ******************************************* PROJECT methods

  static async createProject(projectData) {
    let res = await this.request(`projects/add`, { ...projectData }, 'post');
    return res;
  };

  static async getAllProjects() {
    let res = await this.request(`projects/all`);
    return res;
  };

  static async getNonMatchedProjects() {
    let res = await this.request(`projects/unmatched`);
    return res;
  };

  static async carouselProjects() {
    let res = await this.request(`projects/carousel`);
    return res;
  };

  static async getUserCreatedProjects(username) {
    let res = await this.request(`projects/created/by/${username}`);
    return res;
  };

  static async viewProject(project_id) {
    let res = await this.request(`projects/${project_id}`);
    return res;
  };

  static async editProject(project_id, updateData) {
    let res = await this.request(`projects/${project_id}`, { ...updateData }, 'patch');
    return res;
  };

  static async deleteProject(project_id) {
    let res = await this.request(`projects/${project_id}`, {}, 'delete');
    return res;
  };

};

export default ProjectAPI;