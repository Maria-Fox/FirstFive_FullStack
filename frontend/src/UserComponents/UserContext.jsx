import React from "react";

// Create user context instance to later provide in App.js and use throughout application. Holds auth user

const UserContext = React.createContext();

export default UserContext;