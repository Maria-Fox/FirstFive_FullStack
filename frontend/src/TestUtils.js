// Used to create sample test user data & test UserContect provider..

import React from "react";
import UserContext from "./UserComponents/UserContext";

const authUser = {
  username: "softwareDev1"
};

// sample / data
const matchedProjectIds = [1, 2, 3, 4];

const userWithMatchState = { username: "softwareDev1", matchedProjectIds: [] };

// Creating a context provider and passing in the test user, matching ids. Passing in chidlren.
const UserProvider =
  ({ children, userWithMatchState }) => {
    <UserContext.Provider value={userWithMatchState}>
      {children}
    </UserContext.Provider>
  };

export { UserProvider, authUser, userWithMatchState };