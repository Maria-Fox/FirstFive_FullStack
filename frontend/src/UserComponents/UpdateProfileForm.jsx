import React, { useEffect, useState, useContext } from "react";
import { useParams, useNavigate } from "react-router-dom";
import UserContext from "./UserContext";
import API from "../API";
import AlertNotification from "../Common/AlertNotifications";
// import "./UserProfile.css";
import { Card, Form, FormGroup, Label, Input, Button } from "reactstrap";

const UpdateProfileForm = () => {

  const { authUser } = useContext(UserContext);
  const { username } = useParams();
  const navigate = useNavigate();

  const [formData, setFormData] = useState(null);
  const [formErrors, setFormErrors] = useState(null);

  // ***************************************************************

  useEffect(() => {
    async function confirmPermission() {
      if (authUser !== username) {
        // If not the appropriate user then redirect to that users page.
        navigate(`/users/update/${authUser}`);
      };

      let userData = await API.viewAuthUserProfile(username);

      // Fill form with current data.
      setFormData({
        username: userData.username,
        password: "",
        email: userData.email,
        bio: userData.bio
      });


    };
    confirmPermission();
  }, [setFormData]);

  // ***************************************************************

  const handleSubmit = async (e) => {
    try {
      e.preventDefault();
      await API.editUser(authUser, formData);
      navigate(`/users/${authUser}/profile`);
    } catch (err) {
      setFormErrors(err);
      return;
    }
  }

  // ***************************************************************

  let handleChange = (e) => {
    let { name, value } = e.target;
    setFormData(formData => ({
      ...formData,
      [name]: value
    }))
  };

  // ***************************************************************

  return (
    <div>
      {formErrors ? <AlertNotification messages={formErrors} /> : null}

      <h1 className="text-center text-white pt-2 mt-2"> Update Profile</h1>

      <Card className="container text-center">
        {formData ?
          <Form onSubmit={handleSubmit}>

            <FormGroup>
              <Label htmlFor="username" >Username
                <Input
                  type="text"
                  id="username"
                  value={formData.username}
                  name="username"
                  required
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <FormGroup>
              <Label htmlFor="password" >Password
                <Input
                  type="password"
                  id="password"
                  value={formData.password}
                  name="password"
                  required
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <FormGroup>
              <Label htmlFor="Email" >Email
                <Input
                  type="email"
                  id="email"
                  value={formData.email}
                  name="email"
                  required
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <FormGroup>
              <Label htmlFor="bio" >Bio
                <Input
                  type="text"
                  id="bio"
                  value={formData.bio}
                  name="bio"
                  required
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <Button outline color="info" type="submit" className="m-2">
              Update
            </Button>

          </Form>
          :
          <p>Loading....</p>}

      </Card>
    </div>
  )
};

export default UpdateProfileForm;
