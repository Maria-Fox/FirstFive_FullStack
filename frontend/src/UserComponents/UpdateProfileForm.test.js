import React from "react";
import { MemoryRouter } from "react-router-dom";
import UpdateProfileForm from "./UpdateProfileForm";
import UserContext from "./UserContext";
import { authUser } from "../TestUtils";
import { render, waitFor } from "@testing-library/react";
import API from "../API";
// Mock API
jest.mock("../API")

test("Smoke- Renders form for valid user", function () {

  const authUser = { username: "softwareDev1" };

  render(
    <MemoryRouter>
      <UserContext.Provider value={authUser}>
        <UpdateProfileForm />
      </UserContext.Provider>
    </MemoryRouter>
  );
});

test("Component populates form w current userData", function () {
  // Mock API call & respons data. Retrieving username through params.

  // the component holds the response value as an object (userData) then updates that with that.

  API.mockImplementationOnce(() => {
    return {
      request: () => {
        return { userData: { username: "softwareDev1", email: "email@aol.com", bio: "bio" } };
      },
    };
  });

  const renderedBody = render(
    <UserContext.Provider value={authUser}>
      <MemoryRouter>
        <UpdateProfileForm />
      </MemoryRouter>
    </UserContext.Provider>
  );

  waitFor(() => {
    expect(renderedBody.getByText("email@aol.com")).toBeInTheDocument();
    expect(renderedBody.getByText("OG bio")).toBeInTheDocument();
  });
});