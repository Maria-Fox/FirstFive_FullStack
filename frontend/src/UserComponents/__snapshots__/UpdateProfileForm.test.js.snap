// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading screen for valid user 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-center"
    >
      Update Profile
    </h1>
    <div
      class="container text-center card"
    >
      <p>
        Loading....
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshot - Loading screen for valid user 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-center"
    >
      Update Profile
    </h1>
    <div
      class="container text-center card"
    >
      <p>
        Loading....
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Snapshot- should populate form w current userData 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="text-center"
    >
      Update Profile
    </h1>
    <div
      class="container text-center card"
    >
      <p>
        Loading....
      </p>
    </div>
  </div>
</DocumentFragment>
`;
